server.js

```diff
@@ -1,26 +1,149 @@
+      },
+      message: "Numbers are not allowed"
+    },
+    minlength: 5,
+    maxlength: 140,
+  },
+  hearts: {
+    type: Number,
+    default: 0
+  },
+  createdAt: {
+    type: Date,
+    default: Date.now
+  }
+})
+
+const Thought = mongoose.model('Thought', thoughtSchema)
+
 // Add middlewares to enable cors and json body parsing
 app.use(cors())
-app.use(bodyParser.json())
+app.use(express.json())
 
 // Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send(listEndpoints(app))
+})
+
+app.get('/thoughts', async (req, res) => {
+    const allThoughts = await Thought.find().sort({ createdAt: -1 }).limit(20)
+    res.json(allThoughts)
+  
+})
+
+app.post('/thoughts', async (req, res) => {
+  try {
+    const newThought = await new Thought({message: req.body.message}).save()
+    res.json(newThought)
+  } catch (error) {
+    if (error.code === 11000) {
+      res.status(400).json({ message: 'Duplicated value', fields: error.keyValue })
+    }
+    res.status(400).json(error)
+
+  }
+})
+
+app.post('/thoughts/:id/likes', async (req, res) => {
+  const { id } = req.params
+
+  try {
+    const updatedThought = await Thought.findOneAndUpdate(
+      { 
+        _id: id
+      }, 
+      { 
+        $inc: {
+          hearts: 1 
+        } 
+      }, 
+      {
+         new: true 
+      } 
+    )
+
+    if (updatedThought) {
+      res.json(updatedThought)
+    } else {
+      res.status(404).json({ message: 'Not found'})
+    }
+  } catch (error) {
+    res.status(400).json({ message: 'Invalid request', error })
+  }
+})
+
+app.delete('/thoughts/:id', async (req, res) => {
+  const { id } = req.params 
+
+  try {
+    //v1 - delete
+    // const deletedThought = await Thought.deleteOne({ _id: id })
+    // res.json(deletedThought)
+
+    // v2 - delete
+    const deletedThought = await Thought.findOneAndDelete({ _id: id })
+    if (deletedThought) {
+      res.json(deletedThought)
+    } else {
+      res.status(404).json({ message: 'Not found'})
+    }
+
+  } catch (error) {
+    res.status(400).json({ message: 'invalid request', error})
+  }
+})
+
+app.patch('/thoughts/:id', async (req, res) => {
+  const { id } = req.params
+
+  try {
+    const updatedThought = await Thought.findByIdAndUpdate(id, req.body, { new: true } )
+    if (updatedThought) {
+      res.json(updatedThought)
+    } else {
+      res.status(404).json({ message: 'Not found' })
+    }
+  } catch (error) {
+    res.status(400).json({ message: 'Invalid request', error })
+  }
+})
+
+app.put('/thoughts/:id', async (req, res) => {
+  const { id } = req.params
+
+  try {
+    const updatedThought = await Thought.findOneAndReplace({ _id: id }, req.body, { new: true } )
+    if (updatedThought) {
+      res.json(updatedThought)
+    } else {
+      res.status(404).json({ message: 'Not found' })
+    }
+  } catch (error) {
+    res.status(400).json({ message: 'Invalid request', error })
+  }
```

Comment:
Remember to always restrict access to all endpoints which can manipulate your data stored in your DB.